flash: spi_nor: automatically handle device runtime PM (remove `IDLE_IN_DPD`)

JordanYates opened the pull request:
Automatically handle device runtime PM for all flash API calls. Asynchronously release the device after a small delay to minimise power state transitions under multiple sequential API calls (e.g. NVS).

Remove `SPI_NOR_IDLE_IN_DPD` to simplify the possible transition states in the `spi_nor` driver. This option was originally added 5 years ago when device runtime PM was in a much less mature state.

I do not believe having a separate power management implementation for this one in-tree driver is in the interests of the project.

The behaviour of `SPI_NOR_IDLE_IN_DPD` leads to extremly suboptimal behaviour in use cases where multiple small reads are performed sequentially, see #69588.

Removal of this option does not break the behaviour of any existing applications, only increases current consumption in idle by ~10uA until device runtime PM is enabled.

JordanYates committed:
flash: spi_nor: remove `SPI_NOR_IDLE_IN_DPD`

Remove `SPI_NOR_IDLE_IN_DPD` to simplify the possible transition states
in the `spi_nor` driver. This option was originally added 5 years ago
when device runtime PM was in a much less mature state.

I do not believe having a separate power management implementation for
this one in-tree driver is in the interests of the project.

The behaviour of `SPI_NOR_IDLE_IN_DPD` leads to extremly suboptimal
behaviour in use cases where multiple small reads are performed
sequentially, see #69588.

Removal of this option does not break the behaviour of any existing
applications, only increases current consumption in idle by ~10uA until
device runtime PM is enabled.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
flash: spi_nor: automatically run device runtime PM

Automatically handle device runtime PM for all flash API calls.
Asynchronously release the device after a small delay to minimise power
state transitions under multiple sequential API calls (e.g. NVS).

Signed-off-by: Jordan Yates <jordan@embeint.com>

Laczen commented:
@JordanYates, that is a very nice solution.

de-nordic commented:
@krbvroc1 @lxbrz you have been profiling your designs for power usage recently and are fresh on experience.

de-nordic commented on drivers/flash/spi_nor.c:
```diff
@@ -564,21 +556,19 @@ static void acquire_device(const struct device *dev)
- * mode, and releasing the lock.
- */
+#ifndef CONFIG_SPI_NOR_DEEP_POWER_DOWN_INACTIVITY_TIMEOUT_MS
+#define CONFIG_SPI_NOR_DEEP_POWER_DOWN_INACTIVITY_TIMEOUT_MS 0
```
Do not redefine CONFIG_ when not available, use local identifier.
All c preprocessor definitions should happen on the top of the unit.

de-nordic replied:
Actually it should be impossible to have this undefined, is this even needed?

JordanYates replied:
The symbol `CONFIG_SPI_NOR_DEEP_POWER_DOWN_INACTIVITY_TIMEOUT_MS` depends on `PM_DEVICE_RUNTIME`, so it is undefined when built without PM.
Fixed the initial request

JordanYates commented:
@krish2718 I had to revert #72519 as the flash driver fails to compile on main due to a missing interface driver (`nordic,nrf-exmif`), unless I am missing something obvious:
```
west build -p -b nrf54h20dk/nrf54h20/cpuapp samples/drivers/spi_flash -T sample.drivers.spi.flash
...
zephyr/drivers/flash/libdrivers__flash.a(spi_nor.c.obj):(.rodata.spi_nor_0_config+0x0): undefined reference to `__device_dts_ord_162'
```

krish2718 commented:
> @krish2718 I had to revert #72519 as the flash driver fails to compile on main due to a missing interface driver (`nordic,nrf-exmif`), unless I am missing something obvious:
> 
> ```
> west build -p -b nrf54h20dk/nrf54h20/cpuapp samples/drivers/spi_flash -T sample.drivers.spi.flash
> ...
> zephyr/drivers/flash/libdrivers__flash.a(spi_nor.c.obj):(.rodata.spi_nor_0_config+0x0): undefined reference to `__device_dts_ord_162'
> ```

This means that the driver which defines the device isn't included in the build, let me check and fix this.

krish2718 commented:
> west build -p -b nrf54h20dk/nrf54h20/cpuapp samples/drivers/spi_flash -T sample.drivers.spi.flash

Please go ahead with the revert, this needs more changes from downstream, I had tested this downstream and submitted the PR, haven't verified upstream, thanks.

krish2718 commented:
you also need to revert `b1de9a6c46217b4edc7befb928a9cb2e6d010ca3` same issue as earlier sample, `exmif` peripheral cannot be used in Zephyr now till NCS downstream changes are merged. Do you want to do it in this PR itself? or I can raise a new PR with both reverts, but then you would have to wait till it gets merged and then rebase.

krbvroc1 commented:
> @krbvroc1 you have been profiling your designs for power usage recently and are fresh on experience.

In my case, I found
- The IDLE_IN_DPD was causing the back to back DPD commands. They are sent too fast (14us) and thus violates the tDPDD specification (30us) for the part. That caused the chip to never enter DPD mode.
- I found the SPI bus was not disabled which does not tri-state the pins and caused another 10uA or so.

At this point I settled on setting IDLE_IN_DBD=n and in my application I created the following routing which I current call before and after various calls to write the flash.

```
void dfu_task_enable_flash_spi_bus(bool enable)
{
  const struct device *spi_dev = DEVICE_DT_GET(DT_NODELABEL(spi2));
  const struct device *flash_dev = DEVICE_DT_GET(DT_NODELABEL(mx25r16));

  if (enable)
  {
    pm_device_action_run(spi_dev, PM_DEVICE_ACTION_RESUME);
    pm_device_action_run(flash_dev, PM_DEVICE_ACTION_RESUME);
  }
  else
  {
    pm_device_action_run(flash_dev, PM_DEVICE_ACTION_SUSPEND);
    pm_device_action_run(spi_dev, PM_DEVICE_ACTION_SUSPEND);
  }
}
```
Would using the PM_DEVICE_RUNTIME take care of the SPI bus too? I am thinking not.

How would I continue to enable / disable the SPI bus, which is a significant power consumer, when PM_DEVICE_RUNTIME feature is used? I am having trouble finding it in the documentation now, but I thought I read somewhere not to use PM_DEVICE_RUNTIME while trying to manually change things. Especially if you are introducing a CONFIG_SPI_NOR_DEEP_POWER_DOWN_INACTIVITY_TIMEOUT_MS.

How can you opt in/out devices from PM_DEVICE_RUNTIME. Doesn't this imply all devices in the system support it but maybe I do not want it enabled on all devices.

JordanYates committed:
flash: spi_nor: handle SPI bus power

Automatically request and release the SPI bus before talking to the
flash chip.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates commented:
> Would using the PM_DEVICE_RUNTIME take care of the SPI bus too? I am thinking not.

It does not automatically, but that has been added to this PR.

> but I thought I read somewhere not to use PM_DEVICE_RUNTIME while trying to manually change things.

Correct, we recommend to never use `pm_device_action_run` in application code as it will inevitably break as soon as there are multiple users (a second device on the SPI bus in your example).

> How can you opt in/out devices from PM_DEVICE_RUNTIME. Doesn't this imply all devices in the system support it but maybe I do not want it enabled on all devices.

Device runtime PM has no effect unless it is explicitly enabled for a device. There is no implication about all devices in the application.

JordanYates commented:
> you also need to revert [b1de9a6](https://github.com/zephyrproject-rtos/zephyr/commit/b1de9a6c46217b4edc7befb928a9cb2e6d010ca3) same issue as earlier sample, `exmif` peripheral cannot be used in Zephyr now till NCS downstream changes are merged. Do you want to do it in this PR itself? or I can raise a new PR with both reverts, but then you would have to wait till it gets merged and then rebase.

Happy for you to do a more complete revert in an alternate PR, I don't expect this one will move quickly.

mjchen0 commented on drivers/flash/spi_nor.c:
```diff
@@ -760,11 +767,19 @@ static int spi_nor_read(const struct device *dev, off_t addr, void *dest,
 	release_device(dev);
+
+	/* Release flash power requirement */
+	(void)pm_device_runtime_put_async(dev, K_MSEC(DPD_INACTIVITY_TIMEOUT_MS));
```
If DPD_INACTIVITY_TIMEOUT_MS is 0, should this call pm_device_runtime_put() instead of put_async()? Otherwise it puts the work on the system workqueue without delay, which seems like needless extra work.

JordanYates replied:
This sort of optimization is better suited to occur in the PM implementation so that all users get the benefits and code isn't copy pasted everywhere.

carlescufi commented:
@anangl PTAL

JordanYates commented:
Rebased on `3.7.0` and migration notes added

anangl commented on doc/releases/migration-guide-4.0.rst:
```diff
@@ -99,6 +99,11 @@ Other Subsystems
+ * ``CONFIG_SPI_NOR_IDLE_IN_DPD`` has been removed from the :kconfig:option:`CONFIG_SPI_NOR`
+   driver. An enhanced version of this functionality can be obtained by enabling
+   :ref:`pm-device-runtime` on the device (Tunable with
+   :kconfig:option:`CONFIG_SPI_NOR_ACTIVE_DWELL_MS`).
```
nit: this entry should be added after the runtime PM support and the mentioned Kconfig option are actually added, so in the second commit, not the first one.

de-nordic commented:
@JordanYates This needs rebase as it fails on some unrelated issues.

JordanYates commented:
ping @de-nordic
